### PR TITLE
Changes test assertions to compare wrapped values

### LINKED ISSUE #338: Prefer to assert_eq!(a, Ok(b)) rather than assert_eq!(a.unwrap(), b)
In https://github.com/exercism/rust/pull/292#discussion_r117407845 we decided that asserting equality to an `Ok` gives better error messages than attempting to unwrap (which panics if the observed value is an Err instead of Ok).

We should try to adhere to this.

Find instances of unwrap in asserts by using `git grep 'assert.*unwrap'` (or  `git grep unwrap exercises/*/tests` for a larger hammer) and remove the unwraps by changing the other argument to an `Ok`).

### PR DESCRIPTION
**Pull request by nfiles**
Resolves #338 

Test assertions should compare `Option` or `Result` values directly instead of unwrapping them.

### CODE REVIEW

File: exercises/bowling/tests/bowling.rs [hunk 1]
```diff
@@ -347,7 +347,7 @@ fn the_two_balls_after_a_final_strike_can_not_be_a_non_strike_followed_by_a_stri
 
 #[test]
 #[ignore]
-fn second_bonus_ball_after_a_final_strike_can_not_score_an_invalid_number_of_pins_even_if_first_is_strike() {
+fn second_bonus_ball_after_a_final_strike_can_not_score_an_invalid_number_of_pins_even_if_first_is_strike(){
```
Comment (petertseng) [inline exercises/bowling/tests/bowling.rs]: accidental removal of space?

File: exercises/nucleotide-count/tests/nucleotide-count.rs [hunk 2]
```diff
@@ -9,10 +9,10 @@ fn check_dna(s: &str, pairs: &[(char, usize)]) {
     // check for the presence and value of each key in the given pairs vector.
     let mut m: HashMap<char, usize> = dna::nucleotide_counts(s).unwrap();
     for &(k, v) in pairs.iter() {
-        assert_eq!((k, m.remove(&k).unwrap()), (k, v));
+        assert_eq!((k, m.remove(&k)), (k, Some(v)));
     }
     // may fail with a message that clearly shows all extra pairs in the map
-    assert_eq!(m.iter().collect::<Vec<(&char,&usize)>>(), vec!());
+    assert_eq!(m.iter().collect::<Vec<(&char, &usize)>>(), vec!());
```
Comment (petertseng) [inline exercises/nucleotide-count/tests/nucleotide-count.rs]: the addition of a space is good, but isn't related to removing unwrap, so it should go in a separate commit

**Comment by nfiles** [comment]
I have committed a revised changeset that reverts the spacing issues. :)

**Comment by nfiles** [comment]
No problem!

> After this is merged, are there any remaining instances of unwrap that should be changed?

I think the remaining `unwrap` calls are okay, but I could certainly be wrong. I'm still pretty new to Rust. None of them are being directly used in a test assertion. If the worry is that the `unwrap` call can cause a `panic` at any point in the test (not just the assertion), it might make sense to `assert!(wrapped.is_ok())` before calling unwrap. I'm not sure if that makes sense.

`unwrap` is still called the most in `queen-attack`, `react`, and `triangle`, but none of them are within a test assertion.

1. `queen-attack` - `ChessPosition::new` is unwrapped and passed directly into `Queen::new`.
1. `react` - `reactor.create_compute` is unwrapped because it is passed into other methods to set up the test assertions.
1. `triangle` - `Triangle::build` is unwrapped before the test assertions. An `is_ok` and `is_err` result are both tested.
1. `circular-buffer` - `buffer.read().unwrap().len()` is called within the test assertion, but the unwrapped value is not being directly compared, so I thought it was okay.
1. `nucleotide-count` - `dna::nucleotide_counts(s)` is unwrapped, but the unwrapped `HashMap` is tested with a couple of different iterators, and I don't know how to replace that with a test assertion on a wrapped object.

That's a good point about the `as_slice`/`vec!` instances (`all-your-base` and `variable-length-quantity`). The `example.rs` file was my reference for those and they explicitly returned `Vec` in the optional type.

This pull request resolves #338

**Comment by petertseng** [comment]
Thanks again.

When I am able, I will also check to see whether I agree with your assessment, and assuming I do then it will be time to merge it.

> This pull request resolves #338

That needs to be in the PR description (the one that currently says "Related to Issue #338") OR any commit message in the PR.

**Comment by nfiles** [comment]
Sorry, I understand now. I have updated the PR description.

**Comment by petertseng** [comment]
OK. I agree with all your assessments. And they are definitely not unwraps that are compared, so objective achieved. I may consider a new issue for providing better error messages when the unwraps fail, if there is demand for it (or someone else may open it, that will be sufficient demand). As for this PR, it is time to merge, and close the referenced issue.

I will be making a change to the commit messages. They should be written in imperative form to match with the messages that `git` emits (e.g. on `git merge` and `git revert`), so at a minimum the "Changes" should be "Change"

**Comment by petertseng** [comment]
Thanks (yet) again, great stuff.
